RD-5992 New deployment-update docs (#2369)
cloudify-cosmo/docs.getcloudify.org: content/working_with/manager/update-deployment.md

aliases: /manager/update-deployment/
---

With {{< param product_name >}}, you can update a deployment, which serves multiple purposes:
- Changing the deployment topology: for example, adding a new type of database to an existing deployment of webservers and databases.
- Changing the settings of existing nodes: for example, resizing a volume, or changing the size of a VM instance.
- Updating the real state of the provisioned resources to match the blueprint. For example: after manually resizing a volume by using a cloud provider's console, run update to bring the volume back to the state described in the blueprint.

## Describing a Deployment Update

A deployment update consists of:
- An application blueprint. If a new blueprint is not provided, the current blueprint is used, and the overall deployment topology is kept unchanged.
- New deployment inputs. If new inputs are not provided, existing inputs are used.
- Additional parameters for toggling parts of the update flow on and off.

> The application blueprint is a [yaml blueprint file]({{<relref "developer/blueprints/_index.md">}}), just as any with any other blueprint in {{< param product_name >}} (note that the blueprint represent the desired state of the deployment after the update). Using the example described in the introduction, the updated application blueprint would include a new database type, some new node templates of the new database type, and some new relationships that represent how these new nodes connect to the existing architecture.

Reviewer: ```suggestion
The application blueprint is a [yaml blueprint file]({{<relref "developer/blueprints/_index.md">}}), just as any with any other blueprint in {{< param product_name >}} (note that the blueprint represens the desired state of the deployment after the update). Using the example described in the introduction, the updated application blueprint would include a new database type, some new node templates of the new database type, and some new relationships that represent how these new nodes connect to the existing architecture.
```

##### Detecting configuration drift

In addition to applying the changes based on a new blueprint, the {{< param cfy_manager_name >}} can also:
- check the status of each node instance in the deployment
- if necessary, heal the failing node instances
- check the configuration drift of each node instance - find the differences between the node configuration as described in the blueprint, and the real state of the provisioned resources
- if necessary, update the provisioned resources to match the blueprint.

This status and drift checking is based on the operations defined by each node, usually in a plugin.

## Deployment Update Flow

Updating a deployment is a multi-stage process. The high-level overview of the workflow is:

1. The differences between old and the new blueprint are computed.
1. Steps composing the deployment update are extracted.
1. The {{< param cfy_manager_name >}} data storage is updated with:

>     - the updated deployment attributes (eg. labels)

Reviewer: ```suggestion
    - the updated deployment attributes (e.g. labels)
```

    - the newly-created node instances
    - the updated node properties
    - the updated node instance relationships
1. The `uninstall` operations are executed on each deleted node instance.
1. The `install` operations are executed on each added node instance.
1. The `check_status` (and `heal` if necessary) operations are executed on all started node instances in the deployment.
1. The `check_drift` (and `update` if necessary) operations are executed on all started and healthy node instances in the deployment.
1. The `reinstall` operations are executed on each changed node instance.
1. The deleted nodes and node-instances are removed from the {{< param cfy_manager_name >}} data storage.

For a more in-depth description of these steps, see the [workflow documentation]({{<relref "working_with/workflows/built-in-workflows#the-update-workflow">}})

For notes about implementing the `check_drift` and `update` operations, see [implementing drift check]({{<relref "working_with/manager/update-deployment#implementing-drift-check">}})

## Using {{< param cfy_console_name >}} to Update a Deployment

> To update deployment from the {{< param cfy_console_name >}} go to [Services page]({{<relref "working_with/console/pages/services-page.md">}}), select deployment from the left pane, then click on the

Reviewer: ```suggestion
To update a deployment from the {{< param cfy_console_name >}} go to [Services page]({{<relref "working_with/console/pages/services-page.md">}}), select deployment from the left pane, then click on the
```

> To update deployment from the {{< param cfy_console_name >}} go to [Services page]({{<relref "working_with/console/pages/services-page.md">}}), select deployment from the left pane, then click on the

Reviewer: ```suggestion
To update a deployment from the {{< param cfy_console_name >}} go to [Services page]({{<relref "working_with/console/pages/services-page.md">}}), select **deployment** from the left pane, then click on the
```
The next line should be following here without <enter>

Author: line breaks in the source .md are not rendered in the resulting html

(also btw I note that I only moved this line, didn't actually change it)

If you want to get the same information about update performed in the past:

>  1. Go to **History** tab on specific deployment page and scroll to [Executions widget]({{<relref "working_with/console/widgets/executions.md">}})

Reviewer: ```suggestion
 1. Go to **History** tab on the specific deployment page and scroll to [Executions widget]({{<relref "working_with/console/widgets/executions.md">}})
```

  The overriden input will cause a modification in the `webserver` node (his `port` property was changed). This will trigger an automatic reinstallation of all the instances of the `webserver` node, so the updated port will take affect.
  If the `--skip-reinstall` flag was passed, automatic reinstall will not be triggered, and although the input was overriden to `9090`, the actual port on the existing server will remain `8080`.

> ##### Automatically correcting old inputs' types

Reviewer: ```suggestion
##### Automatically correcting old input types
```
I think? @geokala

Author: nah, `inputs'` is correct  (introduced in #1744)

automatically convert old inputs values' types from `string` to `integer`, `float` or `boolean`,
based on the type of the input declared in a blueprint.

## Implementing drift check

To avoid reinstalling instances when their properties have changed, implement the `check_drift` and `update` operations. Instances of nodes that implement those operations, will run them instead of being reinstalled.

### The `check_drift` operation

> The `cloudify.interfaces.lifecycle.check_drift` operation should examine the node properties, instance runtime properties, and the actual state of the provisioned resource, and compute the differences.

Reviewer: ```suggestion
The `cloudify.interfaces.lifecycle.check_drift` operation should examine the node properties, the instance runtime properties, and the actual state of the provisioned resource, and compute the differences.
```

The instance will be considered drifted if this operation returns a non-empty value (for example, a Python dict describing the differences).

The {{< param cfy_manager_name >}} doesn't inspect the returned value, so it can be any object. Plugin authors are advised to return a description of all the differences, so that the `update` operation can act upon them.

{{% note title="Warning" %}}

> If `check_drift` returns an empty or false value, `update` operations will not run, and even in case of blueprint changes (eg. if the node properties changed), the instances will not be updated or reinstalled. Take care to always return a non-empty value if there are _any_ changes to the instances.

Reviewer: ```suggestion
If `check_drift` returns an empty or false value, `update` operations will not run, and even in case of blueprint changes (e.g. if the node properties have changed), the instances will not be updated or reinstalled. Take care to always return a non-empty value if there are _any_ changes to the instances.
```

{{% /note %}}

> If the `check_drift` operation is not implemented, the instances are only considered drifted if there are relevant blueprint changes (eg. the node properties changed).

Reviewer: ```suggestion
If the `check_drift` operation is not implemented, the instances are only considered drifted if there are relevant blueprint changes (e.g. the node properties have changed).
```

### The `update` operations

The `update` operations are:
- `cloudify.interfaces.lifecycle.update`
- `cloudify.interfaces.lifecycle.update_config`
- `cloudify.interfaces.lifecycle.update_apply`

> Plugin authors can implement any, or all, of those operations. Those three operations can be written in a way to mirror the `create`, `configure`, and `start` operations.

Reviewer: ```suggestion
Plugin authors can implement any, or all, of those operations. Those three operations can be written in a way so that they mirror the `create`, `configure`, and `start` operations.
```

- `cloudify.interfaces.lifecycle.update_config`
- `cloudify.interfaces.lifecycle.update_apply`

> Plugin authors can implement any, or all, of those operations. Those three operations can be written in a way so that they mirror the `create`, `configure`, and `start` operations.

Reviewer 2: This should probably be: "Plugin authors can implement any, or all, of these operations. These three operations can be written so that they mirror the ..."

# deployment update blueprint
description: new_description
```

> ## Known Issues

Reviewer 2: This maybe needs to hang around as it's for a types.yaml, and one from 3.x could still be in use.
